### PR TITLE
Refine ahnet to support changeable psp module

### LINKED ISSUE #1165: Make several layers of AHNet to be optional
**Is your feature request related to a problem? Please describe.**
Currently, the implementation of AHNet aligns with the original paper, which is very heavy.
It may require more than 10 times memory than our UNet.
So I think we can make some deep layers optional, users can select to use them or not with 1 parameter.

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Signed-off-by: yiheng-wang-nv <vennw@nvidia.com>

Fixes #1165 

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv ,

Could you please help train with spleen dataset and see the speed up comparing to full structure version?
Thanks.

**Comment by yiheng-wang-nv** [comment]
> Hi @yiheng-wang-nv ,
> 
> Could you please help train with spleen dataset and see the speed & accuracy comparing to the full structure version?
> Thanks.

Sure

**Review by wyli — APPROVED** [review]
thanks, please confirm the performance and merge.

**Comment by Nic-Ma** [comment]
I am testing the performance on spleen dataset, will update soon.
Thanks.

**Comment by Nic-Ma** [comment]
I tried to train several times in different cases.
The training accuracy is 0.95 which is OK enough, but the training speed is not expected:
I tested with our AMP notebook: https://github.com/Project-MONAI/tutorials/blob/master/acceleration/automatic_mixed_precision.ipynb
Just replaced the UNet with AHNet.

The AMP training is even **slower** than non-AMP training:
On V100 & PyTorch docker 20.10: AMP speed / step = **0.98s**, non-AMP speed / step = **0.87s**.

And to double confirm this issue is not related to HW or SW env, I also changed back to our UNet and tested again:
AMP speed / step = **0.23s**, non-AMP speed / step = **0.39s**.
It's expected and better than PyTorch docker 20.08.

@yiheng-wang-nv could you please help debug deeper on this issue?

Thanks.

**Comment by Nic-Ma** [comment]
Seems the issue is related to the `upsampling` method.
When I changed to use `transpose`: AMP speed / step = **0.54s**, non-AMP speed / step = **0.91s**.
But `transpose` can't support input size as small as `trilinear`(like [96, 96, 96]), need to discuss it.

Thanks.

**Comment by wyli** [comment]
> Seems the issue is related to the `upsampling` method.
> 
> When I changed to use `transpose`: AMP speed / step = **0.54s**, non-AMP speed / step = **0.91s**.
> 
> But `transpose` can't support input size as small as `trilinear`(like [96, 96, 96]), need to discuss it.
> 
> 
> 
> Thanks.
> 
> 

isn't this known from https://github.com/Project-MONAI/MONAI/issues/1023

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Yes, I think it's same as the test results in that link.
Currently, we prefer to use `transpose` in our project and it's deterministic, `trilinear` can't support determinism.
The only problem is that it requires the input size must be 64x, can't support [96, 96, 96].

Thanks.

**Comment by yiheng-wang-nv** [comment]
> Hi @wyli ,
> 
> Yes, I think it's same as the test results in that link.
> Currently, we prefer to use `transpose` in our project and it's deterministic, `trilinear` can't support determinism.
> The only problem is that it requires the input size must be 64x, can't support [96, 96, 96].
> 
> Thanks.

Hi @Nic-Ma @wyli , I've added `nearest` upsample mode, it is a deterministic interpolation method and the size requirement is the same at other linear interpolation modes.
In order to maintain the whole network's determinism, I used the same kernel size as `transpose` upsample mode does for the `maxpool` layer (as shown from line 390 in ahnet.py):
```
        if upsample_mode in ["transpose", "nearest"]:
            """
            To maintain the determinism, the value of kernel_size and stride should be the same.
            (you can check this link for reference: https://github.com/Project-MONAI/MONAI/pull/815 )
            """
            self.maxpool = pool_type(kernel_size=(2, 2, 2)[-spatial_dims:], stride=2)
        else:
            self.maxpool = pool_type(kernel_size=(3, 3, 3)[-spatial_dims:], stride=2, padding=1)
```

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv ,

Thanks for your update.
Using `nearest` mode, AMP speed / step = **0.52s**, non-AMP speed / step = **0.82s**, looks reasonable.
And the loss value of every step in both AMP and non-AMP is deterministic.
Let me try to train on spleen dataset again and test the accuracy with it.

Thanks.

**Review by Nic-Ma — APPROVED** [review]
Looks good to me now.
Thanks.

**Comment by Nic-Ma** [comment]
With `mode=nearest` and `input size=[128, 128, 128]`, `slidingwindow=[192, 192, 192]`, `psp=3`, got: 0.946.
